[OMPIRBuilder] - Make offloading input data persist for deferred target tasks

bhandarkar-pranav opened the pull request:
When we offload to the target, the pointers to data used by the kernel are passed in arrays created by `OMPIRBuilder`. These arrays of pointers are allocated on the stack on the host. This is fine for the most part because absent the `nowait` clause, the default behavior is that target tasks are included tasks. That is, the host is blocked until the offloaded target kernel is done. In turn, this means that the host's stack frame is intact and accessing the array of pointers when offloading is safe. However, when `nowait` is used on the `!$ omp target` instance, then the target task is a deferred task meaning, the generating task on the host does not have to wait for the target kernel to finish. In such cases, it is very likely that the stack frame of the function invoking the target call is wound up thereby leading to memory access errors as shown below.
```
AMDGPU error: Error in hsa_amd_memory_pool_allocate: HSA_STATUS_ERROR_INVALID_ALLOCATION: The requested allocation is not valid.
AMDGPU error: Error in hsa_amd_memory_pool_allocate: HSA_STATUS_ERROR_INVALID_ALLOCATION: The requested allocation is not valid. "PluginInterface" error: Failure to allocate device memory: Failed to allocate from memory manager
fort.cod.out: /llvm/llvm-project/offload/plugins-nextgen/common/src/PluginInterface.cpp:1434: Error llvm::omp::target::plugin::PinnedAllocationMapTy::lockMappedHostBuffer(void *, size_t): Assertion `HstPtr && "Invalid pointer"' failed.
Aborted (core dumped)
```
This PR implements support in OMPIRBuilder to store these arrays of pointers in the task structure that is passed to the target task thereby ensuring it is available to the target task when the target task is eventually scheduled.

bhandarkar-pranav committed:
[OMPIRBuilder] - Make offloading input data persist for deferred target tasks.

When we offload to the target, the pointers to data used by the kernel are
passed in arrays created by OMPIRBuilder. These arrays of pointers are allocated
on the stack on the host. This is fine for the most part because absent the `nowait`
clause, the default behavior is that target tasks are included tasks. That is, the
host waits for the target task (in other words, the target kernel) to complete
before proceeding. In turn, this means that the host's stack frame is intact and
accessing the array of pointers when offloading is safe. However, when nowait is
used on the `!$ omp target` instance, then the target task is a deferred task meaning,
the generating task on the host does not have to wait for the target kernel to finish.
In such cases, it is very likely that the stack frame of the function invoking the target
call is wound up thereby leading to memory access errors as shown below.

AMDGPU error: Error in hsa_amd_memory_pool_allocate: HSA_STATUS_ERROR_INVALID_ALLOCATION: The requested allocation is not valid.
AMDGPU error: Error in hsa_amd_memory_pool_allocate: HSA_STATUS_ERROR_INVALID_ALLOCATION: The requested allocation is not valid.
"PluginInterface" error: Failure to allocate device memory: Failed to allocate from memory manager
fort.cod.out: /llvm/llvm-project/offload/plugins-nextgen/common/src/PluginInterface.cpp:1434:
Error llvm::omp::target::plugin::PinnedAllocationMapTy::lockMappedHostBuffer(void *, size_t):
Assertion `HstPtr && "Invalid pointer"' failed.
Aborted (core dumped)

This PR implements support in OMPIRBuilder to store these arrays of pointers in the task structure
that is passed to the target task thereby ensuring it is available to the target task
when the target task is eventually scheduled.

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7291,15 +7400,56 @@ OpenMPIRBuilder::InsertPointOrErrorTy OpenMPIRBuilder::emitTargetTask(
-    Function *ProxyFn = emitTargetTaskProxyFunction(*this, Builder, StaleCI);
+    // The first argument of StaleCI is always the thread id.
+    // The next few arguments are the pointers to offloading arrays
+    // if any. (See OffloadingArraysToPrivatize)
```
```suggestion
    // if any (see OffloadingArraysToPrivatize).
```

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7291,15 +7400,56 @@ OpenMPIRBuilder::InsertPointOrErrorTy OpenMPIRBuilder::emitTargetTask(
+        createTaskWithPrivatesTy(Task, OffloadingArraysToPrivatize);
+    StructType *PrivatesTy = nullptr;
+
+    if (OffloadingArraysToPrivatize.size())
```
```suggestion
    if (!OffloadingArraysToPrivatize.empty())
```

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7330,17 +7480,19 @@ OpenMPIRBuilder::InsertPointOrErrorTy OpenMPIRBuilder::emitTargetTask(
-        Builder.getInt64(M.getDataLayout().getTypeStoreSize(Task));
+    // plus any other data to be passed to the target task, if any, which
+    // is packed into a struct. kmp_task_t and the struct so created are
+    // packed into a wrapper struct whose type is TaskWithPrivatesTy
```
```suggestion
    // packed into a wrapper struct whose type is TaskWithPrivatesTy.
```

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7378,13 +7530,43 @@ OpenMPIRBuilder::InsertPointOrErrorTy OpenMPIRBuilder::emitTargetTask(
-      Align Alignment = TaskData->getPointerAlignment(M.getDataLayout());
-      Value *TaskShareds = Builder.CreateLoad(VoidPtr, TaskData);
+      Value *Shareds = StaleCI->getArgOperand(SharedArgOperandNo);
+      Value *TaskT = Builder.CreateStructGEP(TaskWithPrivatesTy, TaskData, 0);
+      Value *TaskSharedsPtr = TaskT;
+      if (TaskWithPrivatesTy != Task) {
+        TaskSharedsPtr = Builder.CreateStructGEP(Task, TaskT, 0);
+      }
```
Nit: This logic is duplicated from line 7153. I think it is complex enough to get it refactored into a function. Also, the explanatory comment present in the other place would become accessible when trying to understand the same logic here.

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7378,13 +7530,43 @@ OpenMPIRBuilder::InsertPointOrErrorTy OpenMPIRBuilder::emitTargetTask(
+          Builder.CreateStructGEP(TaskWithPrivatesTy, TaskData, 1);
+      for (unsigned int i = 0; i < OffloadingArraysToPrivatize.size(); ++i) {
+        Value *PtrToPrivatize = OffloadingArraysToPrivatize[i];
+        Type *ArrayType = nullptr;
+        if (auto *GEP = dyn_cast<GetElementPtrInst>(PtrToPrivatize))
+          ArrayType = GEP->getSourceElementType();
+        else if (auto *Alloca = dyn_cast<AllocaInst>(PtrToPrivatize))
+          ArrayType = Alloca->getAllocatedType();
+        else
+          llvm_unreachable("Unhandled Instruction type");
+        assert(ArrayType && "ArrayType cannot be nullptr");
```
Nit: This is duplicated from `createTaskWithPrivatesTy` and I believe it would be a good idea to refactor and reuse. That should make things much easier whenever adding support for the missing instruction types.

skatrak reviewed (commented):
Thank you Pranav for this work, the approach makes sense to me. I have a few comments, but most of them are just small nits.

I think it would be good to check in some of the existing or new Lit tests that the new expected `memcpy` calls and GEP instructions to initialize structures are introduced, and that the right GEP instructions are used to access that data back from within the corresponding outlined functions.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7093,49 +7112,110 @@ static Function *emitTargetTaskProxyFunction(OpenMPIRBuilder &OMPBuilder,
+    else if (auto *Alloca = dyn_cast<AllocaInst>(V))
+      StructFieldTypes.push_back(Alloca->getAllocatedType());
+    else
+      llvm_unreachable("Unhandled Instruction type");
```
All these arrays are created by `OpenMPIRBuilder::emitOffloadingArrays` which only always allocates them on the stack. So, I think they should always be `AllocaInst`. However, since this is a downstream PR that I am upstreaming, this code is a little old and I don't exactly remember why I handeld `GetElementPtrInst` and why I left an `llvm_unreachable` there. If you feel strongly about it, I could take both out and see what breaks ;)

skatrak replied:
Thanks for clearing that up. In that case, I think it would be a good idea to add as a comment to `getOffloadingArrayType` the assumption that passed values must have been produced by `OpenMPIRBuilder::emitOffloadingArrays` and that this means they must be produced by one of the expected instruction types.

bhandarkar-pranav committed:
Address Sergio's comments

bhandarkar-pranav committed:
Clang format changes

bhandarkar-pranav committed:
Update mlir/test/Target/LLVMIR/omptarget-nowait-llvm.mlir

bhandarkar-pranav commented:
@skatrak  - I have updated the PR with a couple of lit tests. I added a new test `omptarget-nowait.mlir` before realizing I could simply update an existing test `omptarget-nowait-llvm.mlir`. Instead of removing the former, I simply updated the latter. If you think I should remove the former, please let me know.

tblah commented:
The approach seems okay to me but this needs detailed review from somebody who understands target offload. Thanks for the patch Pranav.

skatrak reviewed (approved):
Thank you Pranav, LGTM! I just have a small comment to hopefully make things slightly easier to follow.

bhandarkar-pranav committed:
Fix comment

Co-authored-by: Sergio Afonso <safonsof@amd.com>
